File: go_collector/internal/collector/snmp_collector.go
package collector

import (
	"fmt"
	"time"

	"github.com/gosnmp/gosnmp"
)

const (
	oidSysUptime      = "1.3.6.1.2.1.1.3.0"
	oidIfInOctets     = "1.3.6.1.2.1.2.2.1.10.1"
	oidIfOutOctets    = "1.3.6.1.2.1.2.2.1.16.1"
	oidHrStorageUsed  = "1.3.6.1.2.1.25.2.3.1.6.1"
	oidHrStorageSize  = "1.3.6.1.2.1.25.2.3.1.5.1"
	oidHrStorageAlloc = "1.3.6.1.2.1.25.2.3.1.4.1"
	oidHrMemUsed      = "1.3.6.1.2.1.25.2.3.1.6.2"
	oidHrMemTotal     = "1.3.6.1.2.1.25.2.3.1.5.2"
	oidHrMemAllocUnit = "1.3.6.1.2.1.25.2.3.1.4.2"
	oidHrCpuLoad      = "1.3.6.1.2.1.25.3.3.1.2.1"
)

type SNMPCollector struct{}

func NewSNMPCollector() *SNMPCollector { return &SNMPCollector{} }

func (c *SNMPCollector) Collect(target SNMPTarget) (Metrics, error) {
	g := &gosnmp.GoSNMP{
		Target:    target.Host,
		Port:      uint16(target.Port),
		Community: target.Community,
		Version:   gosnmp.Version2c,
		Timeout:   5 * time.Second,
		Retries:   1,
	}
	start := time.Now()
	if err := g.Connect(); err != nil {
		return Metrics{ServiceUp: false}, fmt.Errorf("connect: %w", err)
	}
	defer g.Conn.Close()

	oids := []string{
		oidSysUptime, oidIfInOctets, oidIfOutOctets,
		oidHrStorageUsed, oidHrStorageSize, oidHrStorageAlloc,
		oidHrMemUsed, oidHrMemTotal, oidHrMemAllocUnit,
		oidHrCpuLoad,
	}
	result, err := g.Get(oids)
	latencyMs := int(time.Since(start).Milliseconds())
	if err != nil {
		return Metrics{ServiceUp: false, SNMPLatencyMs: latencyMs},
			fmt.Errorf("get: %w", err)
	}

	m := Metrics{ServiceUp: true, SNMPLatencyMs: latencyMs}
	var memUsedRaw, memTotalRaw, memAlloc int64
	for _, pdu := range result.Variables {
		val := gosnmp.ToBigInt(pdu.Value).Int64()
		switch pdu.Name {
		case "." + oidSysUptime:
			m.UptimeSeconds = val / 100
		case "." + oidIfInOctets:
			m.BandwidthInMB = float64(val) / 1_048_576
		case "." + oidIfOutOctets:
			m.BandwidthOutMB = float64(val) / 1_048_576
		case "." + oidHrCpuLoad:
			m.CPUUsagePercent = float64(val)
		case "." + oidHrMemUsed:
			memUsedRaw = val
		case "." + oidHrMemTotal:
			memTotalRaw = val
		case "." + oidHrMemAllocUnit:
			memAlloc = val
		}
	}
	if memAlloc > 0 {
		m.RAMUsageMB = int((memUsedRaw * memAlloc) / (1024 * 1024))
		m.RAMTotalMB = int((memTotalRaw * memAlloc) / (1024 * 1024))
	} else if memUsedRaw > 0 {
		// fallback: assume 1KiB allocation units (common on Linux net-snmp)
		m.RAMUsageMB = int(memUsedRaw / 1024)
		m.RAMTotalMB = int(memTotalRaw / 1024)
	}

	var used, size, alloc int64
	for _, pdu := range result.Variables {
		v := gosnmp.ToBigInt(pdu.Value).Int64()
		switch pdu.Name {
		case "." + oidHrStorageUsed:
			used = v
		case "." + oidHrStorageSize:
			size = v
		case "." + oidHrStorageAlloc:
			alloc = v
		}
	}
	if size > 0 && alloc > 0 {
		m.DiskUsagePercent = float64(used) / float64(size) * 100
	}
	return m, nil
}

File: node_collector/src/collector/snmpCollector.js
'use strict';

const snmp = require('net-snmp');

const OIDS = {
  sysUptime:      '1.3.6.1.2.1.1.3.0',
  ifInOctets:     '1.3.6.1.2.1.2.2.1.10.1',
  ifOutOctets:    '1.3.6.1.2.1.2.2.1.16.1',
  hrStorageUsed:  '1.3.6.1.2.1.25.2.3.1.6.1',
  hrStorageSize:  '1.3.6.1.2.1.25.2.3.1.5.1',
  hrStorageAlloc: '1.3.6.1.2.1.25.2.3.1.4.1',
  hrMemUsed:      '1.3.6.1.2.1.25.2.3.1.6.2',
  hrMemTotal:     '1.3.6.1.2.1.25.2.3.1.5.2',
  hrMemAllocUnit: '1.3.6.1.2.1.25.2.3.1.4.2',
  hrCpuLoad:      '1.3.6.1.2.1.25.3.3.1.2.1',
};

function collect(target) {
  return new Promise((resolve) => {
    const start = Date.now();

    const session = snmp.createSession(target.host, target.community, {
      port: target.port,
      version: snmp.Version2c,
      timeout: 5000,
      retries: 1,
    });

    session.get(Object.values(OIDS), (error, varbinds) => {
      const latencyMs = Date.now() - start;
      session.close();

      if (error) {
        resolve({ serviceUp: false, snmpLatencyMs: latencyMs });
        return;
      }

      const vals = {};
      for (const vb of varbinds) {
        if (snmp.isVarbindError(vb)) continue;
        const oid = vb.oid.replace(/^\./, '');
        // Counter64 arrives as a Buffer; all others are plain numbers.
        vals[oid] = Buffer.isBuffer(vb.value)
          ? Number(vb.value.readBigUInt64BE(0))
          : Number(vb.value);
      }

      const get = (oid) => vals[oid] ?? 0;

      const memUsed  = get(OIDS.hrMemUsed);
      const memTotal = get(OIDS.hrMemTotal);
      const memAlloc = get(OIDS.hrMemAllocUnit);
      const storUsed  = get(OIDS.hrStorageUsed);
      const storSize  = get(OIDS.hrStorageSize);
      const storAlloc = get(OIDS.hrStorageAlloc);

      let ramUsageMB = 0;
      let ramTotalMB = 0;
      if (memAlloc > 0) {
        ramUsageMB = Math.floor((memUsed * memAlloc) / (1024 * 1024));
        ramTotalMB = Math.floor((memTotal * memAlloc) / (1024 * 1024));
      } else if (memUsed > 0) {
        // fallback: assume 1 KiB allocation units (common on Linux net-snmp)
        ramUsageMB = Math.floor(memUsed / 1024);
        ramTotalMB = Math.floor(memTotal / 1024);
      }

      const diskUsagePercent =
        storSize > 0 && storAlloc > 0
          ? (storUsed / storSize) * 100
          : 0;

      resolve({
        serviceUp:        true,
        snmpLatencyMs:    latencyMs,
        uptimeSeconds:    Math.floor(get(OIDS.sysUptime) / 100),
        bandwidthInMB:    get(OIDS.ifInOctets) / 1_048_576,
        bandwidthOutMB:   get(OIDS.ifOutOctets) / 1_048_576,
        cpuUsagePercent:  get(OIDS.hrCpuLoad),
        ramUsageMB,
        ramTotalMB,
        diskUsagePercent,
      });
    });
  });
}

module.exports = { collect };
